src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1428 @@
+                    or BoundITuplePattern
+                    or BoundConstantPattern
+                    or BoundDeclarationPattern
+                    or BoundDiscardPattern
+                    or BoundTypePattern
+                    or BoundRelationalPattern
+                    or BoundNegatedPattern, $"This rewriter doesn't support pattern {node} yet.");
+
+                return base.Visit(node);
+            }
+
+            // Updates the eval sequence from [...eval sequence...]
+            // to [...eval sequence..., ...eval for left..., ...eval for right..., binaryOperation]
+            // in the general case.
+            //
+            // If the left is a skipped, then update it to
+            // [...eval sequence..., ...eval for right...]
+            //
+            // If the right is a skipped, then update it to
+            // [...eval sequence..., ...eval for left...]
+            //
+            // If both are skipped, then no change occurs.
+            public override BoundNode? VisitBinaryPattern(BoundBinaryPattern node)
+            {
+                bool disjunction = node.Disjunction;
+
+                if (_negated)
+                {
+                    disjunction = !disjunction;
+                }
+
+                var stack = ArrayBuilder<BoundBinaryPattern>.GetInstance();
+
+                BoundBinaryPattern? current = node;
+                do
+                {
+                    stack.Push(current);
+                    current = current.Left as BoundBinaryPattern;
+                }
+                while (current != null && current.Disjunction == node.Disjunction);
+
+                var saveExpectingOperandOfDisjunction = _expectingOperandOfDisjunction;
+                _expectingOperandOfDisjunction = disjunction;
+
+                current = stack.Pop();
+                Debug.Assert((current.Left is BoundBinaryPattern binary && binary.Disjunction != node.Disjunction) || current.Left is not BoundBinaryPattern);
+
+                bool skippedAllLeft = true;
+
+                int startOfLeft = _evalSequence.Count;
+                Visit(current.Left);
+                int endOfLeft = _evalSequence.Count - 1;
+
+                if (startOfLeft <= endOfLeft)
+                {
+                    skippedAllLeft = false;
+                }
+
+                do
+                {
+                    if (skippedAllLeft && stack.IsEmpty)
+                    {
+                        _expectingOperandOfDisjunction = saveExpectingOperandOfDisjunction;
+                    }
+
+                    int startOfRight = _evalSequence.Count;
+                    Visit(current.Right);
+                    int endOfRight = _evalSequence.Count - 1;
+
+                    if (!skippedAllLeft && startOfRight <= endOfRight)
+                    {
+                        PushBinaryOperation(node.Syntax, endOfLeft, disjunction);
+                    }
+
+                    startOfLeft = startOfRight;
```

Comment:
>startOfLeft = startOfRight; [](http://example.com/codeflow?start=20&length=27)

This looks incorrect. In this loop we are building up the left side, adding more and more to it. Therefore, the ```startOfLeft``` doesn't change, only its end is changing.

Reply:
We're doing a stack-based approach, so each time we handle one element from the stack, it becomes the new left and we update all the locals related to the left.

Reply:
Found a compromise by renaming `startOfLeft` to `startOfFirstLeft`.